From 018c888fe6ee1cadc30fe10354396c16ba29449e Mon Sep 17 00:00:00 2001
From: Nadav Hugim <nhugim@habana.ai>
Date: Wed, 12 Jan 2022 12:25:14 +0200
Subject: [PATCH] [SW-62217] preboot: fit: add print func on error

adding new emum for func fit_format_print_error
this func is using in fit_check_format func.

Change-Id: Id22d32a173d16c9fd30a96f896395e3342db6beb
Signed-off-by: Nadav Hugim <nhugim@habana.ai>
---

diff --git a/lib/preboot/fit.c b/lib/preboot/fit.c
index e39e661..8781974 100644
--- a/lib/preboot/fit.c
+++ b/lib/preboot/fit.c
@@ -73,14 +73,6 @@
 	const char *arch;
 };
 
-enum fit_load_result {
-	FIT_LOAD_SUCCESS = 0,
-	FIT_LOAD_CORRUPTED,
-	FIT_LOAD_BAD_LOAD_ADDR,
-	FIT_LOAD_NOT_RUNNABLE,
-	FIT_LOAD_SEC_FAILED,
-};
-
 struct fit_image images[FIT_MAX_IMAGES];
 struct fdt_region fdt_regions[MAX_FDT_REGIONS];
 struct image_region region[MAX_FDT_REGIONS];
@@ -918,6 +910,49 @@
  */
 __weak void preboot_secondary_cpu_stop(void) {}
 
+void fit_load_print_error(enum fit_load_result errtype)
+{
+	switch (errtype) {
+	case FIT_LOAD_BAD_LOAD_ADDR:
+	case FIT_LOAD_NOT_RUNNABLE:
+	case FIT_LOAD_CORRUPTED:
+		preboot_puts("FIT loading failed\n");
+		break;
+	case FIT_LOAD_SEC_FAILED:
+		preboot_puts("FIT security validation failed\n");
+		break;
+	default:
+		preboot_puts_hex("unrecognized error value for: fit_load_result",
+			(unsigned int)errtype, false);
+		break;
+	}
+}
+
+void fit_format_print_error(enum fit_format_result errtype)
+{
+	switch (errtype) {
+	case FIT_FORMAT_HEADER_CHECK_FAILED:
+		preboot_puts("FIT header check failed\n");
+		break;
+	case FIT_FORMAT_NO_DESCRIPTION:
+		preboot_puts("Wrong FIT format: no description\n");
+		break;
+	case FIT_FORMAT_NO_IMAGES_PARENT_NODE:
+		preboot_puts("Wrong FIT format: no images parent node\n");
+		break;
+	case FIT_FORMAT_CURRUPTED_LOAD_TYPE:
+		preboot_puts("Currupted load type\n");
+		break;
+	case FIT_FORMAT_UNSUPPORTED_LOAD_TYPE:
+		preboot_puts("Unsupported load type\n");
+		break;
+	default:
+		preboot_puts_hex("unrecognized error value for: fit_format_result",
+			(unsigned int)errtype, false);
+		break;
+	}
+}
+
 int fit_load_images(void *fit)
 {
 	enum fit_sec_failure sec_res = FIT_SEC_NONE;
diff --git a/lib/preboot/fit.h b/lib/preboot/fit.h
index 5088c38..fa1f484 100644
--- a/lib/preboot/fit.h
+++ b/lib/preboot/fit.h
@@ -8,6 +8,25 @@
 #ifndef _HL_FIT_H_
 #define _HL_FIT_H_
 
+enum fit_load_result {
+	FIT_LOAD_SUCCESS = 0,
+	FIT_LOAD_CORRUPTED,
+	FIT_LOAD_BAD_LOAD_ADDR,
+	FIT_LOAD_NOT_RUNNABLE,
+	FIT_LOAD_SEC_FAILED,
+};
+
+enum fit_format_result {
+	FIT_FORMAT_SUCCESS = 0,
+	FIT_FORMAT_HEADER_CHECK_FAILED,
+	FIT_FORMAT_NO_DESCRIPTION,
+	FIT_FORMAT_NO_IMAGES_PARENT_NODE,
+	FIT_FORMAT_CURRUPTED_LOAD_TYPE,
+	FIT_FORMAT_UNSUPPORTED_LOAD_TYPE,
+};
+
+void fit_load_print_error(enum fit_load_result errtype);
+void fit_format_print_error(enum fit_format_result errtype);
 int fit_load_images(void *fit);
 void fit_run_image(void);
 int fit_validate_size(const void *fit, int len);
